### PR TITLE
fix!: Bump postgres package version.

### PR DESCRIPTION
**Pull request by SandPod**
### Changes:
- Bumps Postgres package to ^3.0.
- Introduce Postgres query mode selection for unsafe queries.
- Refactors the code used to make calls for analyzing the database.

The reason we need to expose the query mode is because the driver now seems to always initiate an extended query behavior for all requests even if no functionality is used. The extended query behavior does not allow multiple queries to be executed as a single query, which is the way we apply our migrations. Therefore we expose it as a configurable for unsafe queries.

Closes: https://github.com/serverpod/serverpod/issues/1766

## Guidance for review:
Best reviewed commit by commit, it is mostly about adapting files to the new behaviors of the interface.

## Pre-launch Checklist

- [ ] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [ ] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [ ] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [ ] All existing and new tests are passing.
- [ ] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

Probably, there are no breaking interface changes for Serverpod but the driver behavior is different compared to the last version which might pose behavior changes for the query interface of our users.

### CODE REVIEW

File: packages/serverpod/lib/src/database/concepts/query_mode.dart [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/// Options for the Query Execution Mode
+enum QueryMode {
+  /// Extended Query Protocol
+  extended,
+
+  /// Simple Query Protocol
+  simple
+}
```
Comment (Isakdl) [inline packages/serverpod/lib/src/database/concepts/query_mode.dart:8]: Question: Why is this needed? What is the difference? 

If we are only using this for the migration stuff wouldn't it be better to have a dedicated method for that instead?

Does this affect the generated code in anyway? Do we have to expose this functionality there too?

File: packages/serverpod/lib/src/database/adapters/postgres/database_connection.dart [hunk 2]
```diff
@@ -378,38 +401,17 @@ class DatabaseConnection {
...
     Transaction? transaction,
+    QueryMode? queryMode,
   }) async {
-    var postgresTransaction = _castToPostgresTransaction(transaction);
-
-    var startTime = DateTime.now();
-    try {
-      var context =
-          postgresTransaction?.executionContext ?? _postgresConnection;
+    var result = await _query(
```
Comment (Isakdl) [inline packages/serverpod/lib/src/database/adapters/postgres/database_connection.dart:406]: Question: if we are passing everything to query, do we need to have the execute method still? 

What is the real difference here between execute and query?

File: packages/serverpod/lib/src/database/adapters/postgres/database_connection.dart [hunk 3]
```diff
@@ -329,33 +333,52 @@ class DatabaseConnection {
...
+    var timeout =
+        timeoutInSeconds != null ? Duration(seconds: timeoutInSeconds) : null;
 
     var startTime = DateTime.now();
     try {
       var context =
           postgresTransaction?.executionContext ?? _postgresConnection;
 
-      var result = await context.query(
+      var result = await context.execute(
```
Comment (Isakdl) [inline packages/serverpod/lib/src/database/adapters/postgres/database_connection.dart:365]: Question: Did they change the functionality of execute / query ?

**Review by Isakdl — COMMENTED** [review]
Some questions on the interface changes.

File: packages/serverpod/lib/src/database/concepts/query_mode.dart (hunk 1 above)
Comment (SandPod) [reply packages/serverpod/lib/src/database/concepts/query_mode.dart:8]: The extended query protocol allows you to supply parameters or create variables in the query, but then it is limited to only a single "statement" performed for each query. We use the extended query protocol when we analyze the database for example, there we create arrays and do calculations in the database as part of the query.

The extended query protocol will also be required if we start using substitutions in Postgres. At the moment we don't use this but it could be used in the future.

The simple query does not allow any substitution or variable declarations but does allow multiple statements to be included in a single query, we use that for applying our migrations.

The driver previously allowed us to always use the extended query protocol but still perform multiple statement. I suspect something is added to the query not that marks it as exclusively extended. Before it used to differentiate between execute and query which means that execute could maybe have been done as a simple query while query was executed as an extended query.

We do need to expose it in the unsafe queries, possibly only in the `execute` method, since our users could need to perform multiple "statements" in a single query, and we default to the extended interface. But we shouldn't need to expose it in our generated code since we control how the query is executed there and we only perform a single query.

File: packages/serverpod/lib/src/database/adapters/postgres/database_connection.dart (hunk 2 above)
Comment (SandPod) [reply packages/serverpod/lib/src/database/adapters/postgres/database_connection.dart:406]: The real different is the return value, execute return the number of rows effected while query returns a result.

File: packages/serverpod/lib/src/database/adapters/postgres/database_connection.dart (hunk 3 above)
Comment (SandPod) [reply packages/serverpod/lib/src/database/adapters/postgres/database_connection.dart:365]: They removed query completely from the interface, only execute exists now.
